### ISSUE TITLE
Crafting talent implementation makes no sense

### ISSUE THREAD

**Issue by cemeceme** [StrangeLoopGames/EcoIssues#21909]
I have looked at a few other reports that discuss similar issues, but wanted to give a more technical description on what is going on.

**Functional issues**
The way talents are implemented is very odd. For one, I would expect talents to be applied either per table or per skill a craft would fall under. This is not the case. Talents apply per recipe ingredient, regardless of the skill that would apply. For instance, this means that a carpentry lavish workspace talent can be made to apply to smelting iron at an anvil, which makes no sense. 

If this is intentional however, there is another issue. Only one talent can apply per ingredient in a craft. So even if the previous point was intentional, it would still mean that a smelting talent cannot apply to a recipe that already has another talent assigned to it. This is both due to recipes ingredients only taking a single talent per ingredient, but also because the `TalentModifiedValue` class only allowing a single talent to be applied.

**Visual issues**
What is worse, is that the tool-tips are inconsistent with the above points too. Even though a talent can only apply to the ingredient that it was defined for, the tool-tip for applying talents list the talent to apply for any `stackable` that matches. This has the effect that if ANY crafting ingredient defines a talent for an item type, ALL crafts that use that item type will show to be affected by that talent, even if they are not. This is because the talents are held in `TalentModifiedValueManager` and are defined and presumably stored as a talent applying to a type and therefore any information on which craft it specifically applies to will be lost.

Another inconsistency to the second functionality point is that talent benefits are shown as a list of applying talents, even though only one can ever apply at once. In fact, only the correct talent can apply even if the intent was to apply only one benefit per craft. This means that you will have some crafts that list an applying talent, that actually modifies the crafting quantities and others that do not, even if some talents are shown to apply. 

**Reproduction**
_scenario 1:_
  Take the Lavish workspace talent for carpentry.
  Check recipes in the carpentry table.
  Check recipes in other types of tables.
 
  EXPECTED:
  All recipes in the carpentry table will gain an efficiency boost (as long as the ingredients are dynamic).
  Either no recipes, or recipes that make sense in other tables will have efficiency gains.

  ACTUAL:
  Some recipes in the carpentry table will not gain an efficiency boost (e.g. Hewn-logs, Basic Upgrade1, Waterwheel) but still show claim to in the tool-tips.
  Recipes in other tables will claim to gain efficiencies from the carpentry talent, but not gain any actual benefits. (e.g. iron bars will show to gain carpentry lavish workspace efficiencies in the anvil)

_scenario 2:_
  Take the Lavish workspace talent the following: Carpentry, Advanced smelting, Basic engineering, pottery.
  Check the recipes in the anvil and look at the tool-tip for any iron bar ingredient.

  EXPECTED:
  No benefits from any talents should be listed (as none of the aforementioned talents should apply).

  ACTUAL:
  The tool-tip claims that all the talents from above would give benefits.
  As the icing on the cake: the cost for the iron bar is not affected in the the slightest, even though it is claimed that the 5% efficiency of 4 different talents apply.

**TLDR list:**
- [ ] Talents apply per recipe, instead of for a skill or table
- [ ] Only one talent may apply per recipe ingredient.
- [ ] Talents are shown to apply for ALL matching item types, instead of the particular craft it is defined for.
- [ ] Talent tool-tips are a list, but only ONE CORRECT talent can apply, regardless of what the tool-tip claims.
- [ ] Talents are shown to apply to crafting ingredients, even though none that are listed actually apply.

Note that the list shows all inconsistencies, but some elements are mutually exclusive.

**Comment by IcedForge** [comment]
Hello, thank you for reporting.  Could you clarify which version this is happening on? Also can you provide your logs as described in the template for making the issue?

**Comment by cemeceme** [comment]
I can guarantee that this happens on the latest version (9.3.4) but I'm fairly certain that this has been the case for a while.
I don't see how logs would help in this case, as it is easily reproducible in a clean install of the server.

While I noticed something was off during some of my past playthroughs on earlier versions, it was only just now that I looked into this issue while writing a mod. However this does not mean that the issue is mod related, just that I was looking closely at how talents are applied that I noticed that something peculiar was going on. What I have written previously specifically outlines what the current implementation is doing after extensive testing. 

In fact, there are multiple other issues here that have reported parts of this one:
#20599 Here, some of the comments talk about how talents don't apply as they are shown in the tool-tip
#12407 General mention of talents not applying properly
#21204 This one is a closed issue, but reports my exact findings, but without the inconsistency of how the effect apply
#20148 Same as above
#16107 Issues in wording

**Comment by IcedForge** [comment]
Sorry for not responding to this earlier but as you summarized in the issues are just that most of this is bugs as detailed and replied to in those issues both in length on how its behaving the way it is but also why as well as potential fixes short term for them.

But the TL;DR is that this post although is a great summary of list of issues it can be considered a duplicate and will be closed as such however we are working on fixing bugs as we continue to develop as well as reshape the talents and finish them off as soon as we can.
